teps/0139-trusted-artifacts.md

```diff
@@ -0,0 +1,477 @@
+                mkdir "$A_FOLDER_PATH"
+                date +%s | tee "${A_FOLDER_PATH}/a.txt"
+                date +%s | tee "${A_FOLDER_PATH}/b.txt"
+                date +%s | tee "${A_FOLDER_PATH}/c.txt"
+      - name: consumer
+        taskSpec:
+          params:
+            - name: aFileArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+            - name: aFolderArtifact
+              type: artifact # inbuilt object schema (path, hash, type)
+          steps:
+            - name: consume-content
+              image: bash:latest
+              script: |
+                #!/usr/bin/env bash
+                # A step is prepended, which will automatically check the hashes
+                # and fail the task with a specific reason if there is no match
+                # The artifact is copied to local disk before the hash is checked
+                # to protect from changes on the workspace and support the case
+                # of a consumer with write access to the disk.
+                # Do something with the verified content. 
+                # The path from the object params corresponds to the result's "data.path"
+                # The data.path variable corresponds to the local folder
+                # and resolves to a path on the workspace
+                echo "File content"
+                cat $(params.aFileArtifact.data.path)
+                echo "Folder content"
+                find $(params.aFolderArtifact.data.path) -type f
+        params:
+          - name: aFileArtifact
+            value: $(tasks.producer.results.aFileArtifact)
+          - name: aFolderArtifact
+            value: $(tasks.producer.results.aFolderArtifact)
+  workspaces:
+  - name: artifactStorage
+    volumeClaimTemplate:
+      spec:
+        accessModes:
+          - ReadWriteOnce
+        resources:
+          requests:
+            storage: 1Gi
+```
+
+</details>
+
+</br>
+
+### Notes and Caveats
+
+Some questions raised during the initial presentation:
+
+* Q: Can we restrict access to the persistent volumes to the injected steps? Maybe using TEP-0029
+* A: We could mount the workspace to injected steps instead of relying on propagated workspaces. This would not prevent users from mounting the workspace to other steps / sidecars though, unless we add validation to prevent that. However that would mean that a consumer could not use the artifact workspace to produce another artifact, which would be problematic
+
+* Q: Using an emptyDir secures the data, but may be less performant than writing directly
+* A: On the producing side, we need to let users write to an `emptyDir`, and the injected step will calculate the hash and then copy the data to the workspace. On the consuming side, we need to copy data to an `emptyDir` and then verify the checksum, or else we cannot be sure that the data has not been compromised after the checksum verification
```

Comment:
could you explain more about how emptyDir ensures data has not been compromised after verification? Another step within the same task could modify things in the emptyDir right? And on the PVC, unless it is read write many, two pods won't be able to simultaneously write?

Reply:
cc @pritidesai this is the question I raised at the data interface wg, might be helpful for you to comment here as well

Reply:
@dibyom steps within a task are executed sequentially. 

Another step could alter executed after the first could alter what the first step produced, but this proposal only ensures that a consuming `Task2` can verify that what it receives as the output of `Task1` is what `Task1` actually produced. It doesn't really tell `Task2` anything about the status in between steps.

A workload that belongs or does not belong to the `PipelineRun` could alter the content of the workspace in between tasks, but after this TEP such change would invalidate the provenance data, which is written in the `TaskRun` status and signed (once we have Spire integration). If a malicious third-party workload tried to update the provenance in the status, the status signature would be invalidated.

About the "read write many", the "many" refers to (thanks @pritidesai for pointing that out) "many nodes" i.e. all `Pods` on the same nodes will have access to the PVC.